File: app/src/main/java/com/example/narthana/genpass/GraphemeLengthFilter.java
package com.example.narthana.genpass;

import android.text.InputFilter;
import android.text.Spanned;

import com.ibm.icu.text.BreakIterator;

/**
 * Created by narthana on 25/01/17.
 */

class GraphemeLengthFilter implements InputFilter
{
    private int mMax;

    GraphemeLengthFilter(int max) { mMax = max; }

    public int getMax() { return mMax; }

    // We are replacing the range dstart...dend of dest with the rance start...end of source
    // The return value is the replacement or null to use source unadulterated
    @Override
    public CharSequence filter(CharSequence source, int start, int end,
                               Spanned dest, int dstart, int dend)
    {
        //  number of graphemes we must keep
        int keep = mMax - (graphemeCount(dest, 0, dest.length())
                           - graphemeCount(dest, dstart, dend));

        if (keep <= 0) return "";
        else if (keep >= graphemeCount(source, start, end)) return null; // keep original
        else
        {
            BreakIterator it = BreakIterator.getCharacterInstance();
            it.setText(source.subSequence(start, end).toString());
            int endToKeep = 0;
            while (keep > 0 && (endToKeep = it.next()) != BreakIterator.DONE) --keep;
            return source.subSequence(start, start + endToKeep);
        }
    }

    private int graphemeCount(CharSequence seq, int beginIndex, int endIndex)
    {
        BreakIterator it = BreakIterator.getCharacterInstance();
        it.setText(seq.subSequence(beginIndex, endIndex).toString());
        int count = 0;
        while (it.next() != BreakIterator.DONE) ++count;
        return count;
    }
}

File: app/src/main/java/com/example/android/genpass/text/GraphemeLengthFilter.kt
package com.example.android.genpass.text

import android.text.InputFilter
import android.text.Spanned
import com.ibm.icu.text.BreakIterator
import java.util.*

/**
 * Created by narthana on 25/01/17.
 */

class GraphemeLengthFilter @JvmOverloads
constructor(val max: Int, private val locale: Locale = Locale.getDefault()): InputFilter {
    /**
     * We are replacing the range [dstart]...[dend] of [dest] with the range [start]...[end]
     * of [source]. The return value is the replacement or null to use source unadulterated
     */
    override fun filter(source: CharSequence, start: Int, end: Int,
                        dest: Spanned, dstart: Int, dend: Int): CharSequence? {
        //  keep = number of graphemes we must keep
        val keep = max - (graphemeCount(dest, 0, dest.length) - graphemeCount(dest, dstart, dend))

        return if (keep <= 0) ""
        else if (keep >= graphemeCount(source, start, end)) null // keep original
        else {
            // 0 < keep < graphemeCount(source, start, end).
            // so move as far into source as possible
            BreakIterator.getCharacterInstance(locale).run {
                setText(source.subSequence(start, end).toString())
                // iterate though "keep" graphemes, noting the index at which we end up
                val keepOffset = (1 .. keep).fold(0) { _, _ -> next() }
                source.subSequence(start, start + keepOffset)
            }
        }
    }

    /**
     * Counts the number of graphemes in [seq] from [beginIndex] to [endIndex]
     */
    private fun graphemeCount(seq: CharSequence, beginIndex: Int, endIndex: Int): Int
            = BreakIterator.getCharacterInstance(locale).run {
        setText(seq.subSequence(beginIndex, endIndex).toString())
        var count = 0
        while (next() != BreakIterator.DONE) ++count
        count
    }
}